packages/shield-controller/src/polling-with-timeout-abort.ts

```diff
@@ -0,0 +1,181 @@
+export type RequestEntry = {
+  abortController: AbortController; // The abort controller for the request
+  abortHandler: () => void; // The abort handler for the request
+  timerId: NodeJS.Timeout; // The timer ID for the request timeout
+};
+
+export type RequestFn<ReturnType> = (
+  signal: AbortSignal,
+) => Promise<ReturnType>;
+
+export class PollingWithTimeoutAndAbort {
+  readonly ABORT_REASON_TIMEOUT = 'Timeout';
+
+  readonly ABORT_REASON_USER_CANCELLATION = 'User cancellation';
+
+  // Map of request ID to request entry
+  readonly #requestEntries: Map<string, RequestEntry> = new Map();
+
+  readonly #timeout: number;
+
+  readonly #pollInterval: number;
+
+  constructor(config: { timeout: number; pollInterval: number }) {
+    this.#timeout = config.timeout;
+    this.#pollInterval = config.pollInterval;
+  }
+
+  async pollRequest<ReturnType>(
+    requestId: string,
+    requestFn: RequestFn<ReturnType>,
+    pollingOptions: {
+      timeout?: number;
+      pollInterval?: number;
+      fnName?: string;
+    } = {
+      fnName: '',
+    },
+  ) {
+    const timeout = pollingOptions.timeout ?? this.#timeout;
+    const pollInterval = pollingOptions.pollInterval ?? this.#pollInterval;
+
+    // clean up the request entry if it exists
+    this.abortPendingRequests(requestId);
+
+    // insert the request entry for the next polling cycle
+    const { abortController } = this.#insertRequestEntry(requestId, timeout);
```

Comment:
The whole functionality of this is a bit difficult to grasp as it's spread across the whole file and different handlers (especially the different cleanup handlers are confusing).

Would it be possible to streamline the logic a bit:
1. Create 1 abort controller. If this aborts, the polling aborts and the abort handling is cleaned up.
2. Attached the timeout event to the abort controller. (i.e., on timeout, call `controller.abort()`)